apple/Sources/Sargon/Util/SargonObjectCodable.swift

```diff
@@ -0,0 +1,23 @@
+import Foundation
+import AnyCodable
+
+public protocol SargonObjectCodable: Codable {
```

Comment:
Nope, needed, to disambiguate with some other "profile Atoms" (types) which we **also** make Swift Codable, but where they can be represented as a literal, eg AccountAddress is turned into Codable using singleValeContainer and encode/decode as String.

The name is a hint so that we don't think to use the SargonObjectCodable on a literal repeesentable type

Reply:
Not sure why it matters that they can be represented as literals, I mean that is true for some Codable, but not all. But to be honest I am not sure about the purpose of the protocol in the first place...

Reply:
We cannot simply add `Codable` constraint on our types in Post Processing of UniFFI bindgen - since our JSON scheme is more advanced that that. So we MUST rely fully on the JSON impl of Profile within Rust. Meaning the only thing we can bridge from Rust to Swift is either JSON as `Data` or JSON as `String`. 

The purpose of the protocol is to be able to bridge JSON `Data` <> Swift `Codable`, for MANY Types. Which I have successfully done :) I'm merely asking, can it be improved....

Reply:
I am not explaining my suggestion very clearly

Reply:
I mean wrapper methods like `sargonEncode` here:
```Swift
struct PhoneInfo { }

func newDeviceInfoFromJsonBytes(jsonBytes: Data) throws -> PhoneInfo {
	fatalError()
}

func deviceInfoToJsonBytes(deviceInfo: PhoneInfo) -> Data {
	fatalError()
}

extension Encoder {
	func sargonEncode(data: Data) throws {
		let dict = try JSONSerialization.jsonObject(with: data, options: []) as! NSDictionary
		let anyEncodable = AnyEncodable(dict)
		var container = singleValueContainer()
		try container.encode(anyEncodable)
	}
}

extension PhoneInfo: Encodable {
	public func encode(to encoder: any Encoder) throws {
		try encoder.sargonEncode(data: deviceInfoToJsonBytes(deviceInfo: self))
	}
}
```
Now all you have to do is implement `encode` like just above.

Reply:
And the other direction:
```Swift
// Do for each type
extension PhoneInfo: Decodable {
	public init(from decoder: any Decoder) throws {
		self = try decoder.sargonDecode(with: newDeviceInfoFromJsonBytes)
	}
}

// Generic helper:
extension Decoder {
	func sargonDecode<T: Decodable>(with function: (Data) throws -> T) throws -> T {
		let container = try singleValueContainer()
		let anyDecodable = try container.decode(AnyDecodable.self)
		let jsonData = try JSONSerialization.data(withJSONObject: anyDecodable.value)
		return try function(jsonData)
	}
}
```

Reply:
```Swift
extension DeviceInfo: SargonObjectCodable {}
```
Surely this is not enough? You also need to:

```Swift
	public init(jsonData: some DataProtocol) throws {
		self = try newDeviceInfoFromJsonBytes(jsonBytes: Data(jsonData))
	}

	public func jsonData() -> Data {
		deviceInfoToJsonBytes(deviceInfo: self)
	}

```

Reply:
Right, but the init and the func bridging to the global scope functions I should have, because when/if UniFFI actually DO support methods on Record (structs) THEN we can delete those. 

So the project structure is quite specific. All wrapped functions are out in a special folder, which is now in total many many hundreds of lines of code. Which we might be able to delete, the whole folder that is. 

So that is why this setup is best.

Reply:
To me it seems that the two solutions in my previous message are the same in that regard.

Reply:
Your solution requires `N` many implementations of encode func and as many of init:decoder, I think that is inferior. For the cost of having a protocol.

Swift is protocol driven, why the discrimination against protocol? :)

Reply:
And I'm saying we MUST have `init:data` and `jsonData()` for now, and we MIGHT be able to remove it later. why we dont wanna *conflate* that with the actual `Codable` impl. Those SHOULD be kept distinct, allowing us later to remove the whole `apple/Sources/Sargon/Extensions/Methods/` folder.

Operating under that constraint: then using a protocol is best. Leaving it at that, let me know if you can come up with anything better than:

```swift
	let dict = try JSONSerialization.jsonObject(with: self.jsonData(), options: []) as! NSDictionary
	let anyEncodable = AnyEncodable(dict)
	var container = encoder.singleValueContainer()
	try container.encode(anyEncodable)
```

and 

```swift
	let container = try decoder.singleValueContainer()
	let anyDecodable = try container.decode(AnyDecodable.self)
	let jsonData = try JSONSerialization.data(withJSONObject: anyDecodable.value)
	try self.init(jsonData: jsonData)
```

Reply:
Aha, so when you talk about global functions you are referring to `newDeviceInfoFromJsonBytes` etc, not my methods which are not more global than yours. So the thing is that you will want to use `init(jsonData:)` in other situations, rather than having to expose `newDeviceInfoFromJsonBytes`. But in my solution, we have `init(decoder:)` instead, which you can't use in those contexts.